### PR TITLE
Feature/re add perfdata pie charts in list views

### LINKED ISSUE #160: Re add perfdata pie charts in list views
In the new lists the pie charts for the perfdata that used to be on the right side of some service list items moved to the detail view. See:

image

vs

image

From feedback from the community I gathered, that those were very useful in the overview and comparatively unneccesary in the detail view :)

### PR DESCRIPTION
**Pull request by flourish86**
resolves #160

### CODE REVIEW

File: library/Icingadb/Common/ListItemDetailedLayout.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+/* Icinga DB Web | (c) 2020 Icinga GmbH | GPLv2 */
```
Comment (nilmerg) [inline library/Icingadb/Common/ListItemDetailedLayout.php:3]: Isn't it 2021?

File: library/Icingadb/Util/ThresholdRange.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
+<?php
+
+/* Icinga Web 2 | (c) 2016 Icinga Development Team | GPLv2+ */
```
Comment (nilmerg) [inline library/Icingadb/Util/ThresholdRange.php:3]: Please adjust accordingly to the project.

File: library/Icingadb/Widget/Detail/ObjectDetail.php [hunk 3]
```diff
@@ -300,22 +300,12 @@ protected function createNotifications()
...
-        $helper->view = Icinga::app()->getViewRenderer()->view;
-
         $content[] = Html::tag('h2', t('Performance Data'));
 
         if (empty($this->object->state->performance_data)) {
             $content[] = new EmptyState(t('No performance data available.'));
         } else {
-            $content[] = new HtmlElement(
-                'div',
-                ['id' => 'check-perfdata-' . $this->object->checkcommand],
```
Comment (nilmerg) [inline library/Icingadb/Widget/Detail/ObjectDetail.php:316]: Please re-add the id. It's necessary for the collapsible state to persist correctly.

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 4]
```diff
@@ -4,11 +4,75 @@
 
 namespace Icinga\Module\Icingadb\Widget;
 
+use Icinga\Application\Icinga;
+use Icinga\Module\Icingadb\Common\ListItemDetailedLayout;
 use Icinga\Module\Icingadb\Compat\CompatPluginOutput;
+use Icinga\Module\Icingadb\Util\Perfdata;
+use Icinga\Module\Icingadb\Util\PerfdataSet;
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:11]: Please remove unused use statements.

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 5]
```diff
@@ -4,11 +4,75 @@
...
+    protected function getStateBallSize()
+    {
+        return StateBall::SIZE_LARGE;
+    }
+
+    protected function assembleFooter(BaseHtmlElement $footer)
+    {
+        $this->pie = HtmlString::create('');
+
+        $statusIcons = Html::tag('div', ['class' => 'status-icons']);
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:33]: Please use `new HtmlElement('div', ['class' => 'status-icons'])` instead. The `Html` class shouldn't be used inside of widgets. (It seems to be a small optimization, but it makes a difference if it's called many times in a row. `Html::tag` is a detour after all.)

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 6]
```diff
@@ -4,11 +4,75 @@
...
+
+        // ToDo(fs): Get `has_comments` from database
+        if ($this->item->comment->limit(1)->execute()->hasResult()) {
+            $statusIcons->add(new Icon('comments', ['title' => t('This item has been commented')]));
+        }
+
+        if (!$this->item->notifications_enabled) {
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:40]: Please add a space between `!` and what follows. There's another occurrence below.

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 7]
```diff
@@ -4,11 +4,75 @@
...
+        $performanceData = Html::tag('div', ['class' => 'performance-data']);
+        if ($this->item->state->performance_data) {
+            $pieChartData = PerfdataSet::fromString($this->item->state->performance_data)->asArray();
+
+            $pies = [];
+            foreach ($pieChartData as $i => $perfdata) {
+                if ($perfdata->isVisualizable()) {
+                    $pies[] = $perfdata->asInlinePie()->render();
+                }
+                // Check if number of visualizable pie charts is larger than PIE_CHART_LIMIT
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:57]: Please add an empty line after closing braces.

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 8]
```diff
@@ -4,11 +4,75 @@
...
+                if (count($pies) > ServiceListItemDetailed::PIE_CHART_LIMIT) {
+                    break;
+                }
+            }
+
+            foreach ($pies as $i => $pie) {
+                if ($i > 3 && count($pies) > ServiceListItemDetailed::PIE_CHART_LIMIT) {
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:64]: What's the `$i > 3` for?

File: library/Icingadb/Widget/PerfdataTable.php [hunk 9]
```diff
@@ -0,0 +1,137 @@
+<?php
+
+/* Icinga Web 2 | (c) 2013 Icinga Development Team | GPLv2+ */
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:3]: Please adjust accordingly to the project.

File: library/Icingadb/Widget/PerfdataTable.php [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+     *
+     * @param   string  $perfdataStr    The perfdata string
+     * @param   bool    $compact        Whether to display the perfdata in compact mode
+     * @param   int     $limit          Max labels to show; 0 for no limit
+     * @param   string  $color          The color indicating the perfdata state
+     *
+     * @return string
+     */
+
+
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:41]: Please remove these empty lines.

File: library/Icingadb/Widget/PerfdataTable.php [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            array(
+                '',
+                t('Label'),
+                t('Value'),
+                t('Min'),
+                t('Max'),
+                t('Warning'),
+                t('Critical')
+            )
+        );
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:71]: Please use brackets for array notations.

File: library/Icingadb/Widget/PerfdataTable.php [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        foreach ($pieChartData as $perfdata) {
+            if ($perfdata->isVisualizable()) {
+                $columns[''] = '';
+                $this->containsSparkline = true;
+            }
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:76]: Empty lines, after closing braces, please :)

File: library/Icingadb/Widget/PerfdataTable.php [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+                    $column === 'min' && floatval($value) === 0.0 ||
+                    $column === 'max' && $perfdata->isPercentage() && floatval($value) === 100
+                ) {
+                    continue;
+                }
+                $columns[$column] = $labels[$column];
+            }
+        }
+
+        if (!$this->containsSparkline) {
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:89]: A space after `!` ;)

File: library/Icingadb/Widget/PerfdataTable.php [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+            $keys = array_slice($keys, 1, -1);
+        }
+
+        $headerRow = Html::tag('tr');
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:93]: `Html::tag` vs `new HtmlElement()`, you know?

File: library/Icingadb/Widget/PerfdataTable.php [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+                            $text
+                        ),
+                        [ 'class' => ($column == 'label' ? 'title' : null) ]
+                    );
+                }
+
+                $this->add(Table::tr([$cols]));
+            }
+
+            continue;
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:134]: That's redundant. Please remove.

File: library/Icingadb/Widget/StateListItem.php [hunk 16]
```diff
@@ -83,6 +83,15 @@ protected function assembleVisual(BaseHtmlElement $visual)
         }
     }
 
+    protected function createFooter()
+    {
+        $footer = Html::tag('div', ['class' => 'footer']);
```
Comment (nilmerg) [inline library/Icingadb/Widget/StateListItem.php:88]:
```suggestion
        $footer = new HtmlElement('div', ['class' => 'footer']);
```

File: public/css/common.less [hunk 17]
```diff
@@ -156,6 +156,46 @@ div.show-more {
...
+    font-weight: bold;
+  }
+
+  .sparkline-col {
+    min-width: 2em;
+    width: 2em;
+    padding: 0;
+  }
+}
+
```
Comment (nilmerg) [inline public/css/common.less:198]: Please put this into its own file in the widgets folder.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Icingadb/Util/PerfData.php [hunk 18]
```diff
@@ -0,0 +1,546 @@
...
+
+        return $pieChart;
+    }
+
+    /**
+     * Format the given value depending on the currently used unit
+     */
+    protected function format($value)
+    {
+        if ($value instanceof ThresholdRange) {
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfData.php:436]: Please add the same check as in https://github.com/Icinga/icingaweb2/pull/4351. Empty values should show `-`, not `0.00`.

File: library/Icingadb/Widget/HostListItemDetailed.php (hunk 5 above)
Comment (nilmerg) [reply library/Icingadb/Widget/HostListItemDetailed.php:32]: That's not resolved, or am i blind?

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 19]
```diff
@@ -4,11 +4,78 @@
...
+
+                // Check if number of visualizable pie charts is larger than PIE_CHART_LIMIT
+                if (count($pies) > ServiceListItemDetailed::PIE_CHART_LIMIT) {
+                    break;
+                }
+            }
+
+            foreach ($pies as $i => $pie) {
+                if (
+                    $i > ServiceListItemDetailed::PIE_CHART_LIMIT - 2
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:65]: Why `- 2`? :joy: If you want to break out after the next pie once the limit is reached, should `$i > ServiceListItemDetailed::PIE_CHART_LIMIT` not suffice?

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 20]
```diff
@@ -4,11 +4,78 @@
...
+                    && count($pies) > ServiceListItemDetailed::PIE_CHART_LIMIT
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:66]: Please remove that. It doesn't make much sense to me to calculate the count on very iteration step. It's also unnecessary in my opinion, the first check should suffice.

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 21]
```diff
@@ -4,11 +4,78 @@
...
+                ) {
+                    $performanceData->add(new HtmlElement('span', [], '…'));
+                    break;
+                }
+                $performanceData->add(HtmlString::create($pie));
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:71]: An empty line after every closing brace :wink:

File: library/Icingadb/Widget/PerfdataTable.php (hunk 10 above)
Comment (nilmerg) [reply library/Icingadb/Widget/PerfdataTable.php:42]: Again, not solved.

File: library/Icingadb/Widget/PerfdataTable.php [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+
+        $headerRow = new HtmlElement('tr');
+        foreach ($keys as $col) {
+            if (isset($col)) {
+                $headerRow->add(new HtmlElement('th', [
+                    'class' => ($col == 'label' ? 'title' : null)
+                ], $labels[$col]));
+            }
+        }
+        $this->getHeader()->add($headerRow);
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:103]: An empty line after...?

File: library/Icingadb/Widget/ServiceListItemDetailed.php [hunk 23]
```diff
@@ -4,11 +4,78 @@
...
+                // Check if number of visualizable pie charts is larger than PIE_CHART_LIMIT
+                if (count($pies) > ServiceListItemDetailed::PIE_CHART_LIMIT) {
+                    break;
+                }
+            }
+
+            foreach ($pies as $i => $pie) {
+                if (
+                    $i > ServiceListItemDetailed::PIE_CHART_LIMIT - 2
+                    && count($pies) > ServiceListItemDetailed::PIE_CHART_LIMIT
```
Comment (nilmerg) [inline library/Icingadb/Widget/ServiceListItemDetailed.php:65]: Just like in class `HostListItemDetailed`, please.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Icingadb/Widget/HostListItemDetailed.php (hunk 19 above)
Comment (flourish86) [reply library/Icingadb/Widget/HostListItemDetailed.php:65]: As discussed, I added a comment there for clarification

File: library/Icingadb/Widget/PerfdataTable.php [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+     * Display the given perfdata string to the user
+     *
+     * @param   string  $perfdataStr    The perfdata string
+     * @param   bool    $compact        Whether to display the perfdata in compact mode
+     * @param   int     $limit          Max labels to show; 0 for no limit
+     * @param   string  $color          The color indicating the perfdata state
+     *
+     * @return string
+     */
+
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:41]: There's this empty line still. Remove please.

File: library/Icingadb/Widget/HostListItemDetailed.php [hunk 25]
```diff
@@ -4,11 +4,81 @@
...
+use ipl\Html\HtmlString;
+use ipl\Web\Widget\Icon;
+use ipl\Web\Widget\StateBall;
 
-class HostListItemDetailed extends HostListItem
+class HostListItemDetailed extends BaseHostListItem
 {
+    use ListItemDetailedLayout;
+
+    protected $pie;
```
Comment (nilmerg) [inline library/Icingadb/Widget/HostListItemDetailed.php:21]: Please add a doc block here.

File: library/Icingadb/Widget/PerfdataTable.php [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+use Icinga\Module\Icingadb\Util\Perfdata;
+use Icinga\Module\Icingadb\Util\PerfdataSet;
+use ipl\Html\Html;
+use ipl\Html\HtmlElement;
+use ipl\Html\HtmlString;
+use ipl\Html\Table;
+
+class PerfdataTable extends Table
+{
+    private $containsSparkline = false;
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:16]: Why is this private? protected should suffice, just like for the other attributes.

And please add a docblock, also to all other attributes you're introducing.

File: library/Icingadb/Widget/PerfdataTable.php [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+
+            foreach ($perfdata->toArray() as $column => $value) {
+                if (
+                    empty($value) ||
+                    $column === 'min' && floatval($value) === 0.0 ||
+                    $column === 'max' && $perfdata->isPercentage() && floatval($value) === 100
+                ) {
+                    continue;
+                }
+                $columns[$column] = $labels[$column];
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:86]: Empty lines after braces...

File: public/css/widget/performance-data-table.less [hunk 28]
```diff
@@ -0,0 +1,41 @@
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2 */
+
+.performance-data-table {
+  width: 100%;
+
+  td {
+    text-align: right;
+    .text-ellipsis;
```
Comment (nilmerg) [inline public/css/widget/performance-data-table.less:8]: Please add parentheses to mixin calls. `.text-ellipsis()`

**Review by nilmerg — CHANGES_REQUESTED** [review]
There are also still changes you didn't address yet.

File: library/Icingadb/Widget/HostListItemDetailed.php (hunk 25 above)
Comment (flourish86) [reply library/Icingadb/Widget/HostListItemDetailed.php:21]: Seems like it wasn't used anyway, so I removed it completely

File: library/Icingadb/Widget/PerfdataTable.php [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+
+use Icinga\Module\Icingadb\Util\Perfdata;
+use Icinga\Module\Icingadb\Util\PerfdataSet;
+use ipl\Html\HtmlElement;
+use ipl\Html\HtmlString;
+use ipl\Html\Table;
+
+class PerfdataTable extends Table
+{
+
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:15]: Where do all these empty lines come from? Please remove.

File: library/Icingadb/Widget/PerfdataTable.php [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+        'data-visible-rows' => 6
+    ];
+
+    protected $perfdataStr;
+
+    protected $compact;
+
+    protected $limit;
+
+    protected $color;
```
Comment (nilmerg) [inline library/Icingadb/Widget/PerfdataTable.php:32]: A doc block please for all other attributes you're introducing, I said. :man_teacher:

**Review by nilmerg — CHANGES_REQUESTED** [review]
You should take another look at class `PerfdataTable`, and one or two at the still not resolved comments.

File: library/Icingadb/Util/PerfData.php [hunk 31]
```diff
@@ -0,0 +1,550 @@
...
+
+namespace Icinga\Module\Icingadb\Util;
+
+use Icinga\Exception\ProgrammingError;
+use Icinga\Module\Monitoring\Object\Service;
+use Icinga\Util\Format;
+use Icinga\Web\Widget\Chart\InlinePie;
+use InvalidArgumentException;
+
+class Perfdata
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfData.php:13]:
```suggestion
class PerfData
```

File: library/Icingadb/Util/PerfDataSet.php [hunk 32]
```diff
@@ -0,0 +1,144 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2+ */
+
+namespace Icinga\Module\Icingadb\Util;
+
+use ArrayIterator;
+use IteratorAggregate;
+
+class PerfdataSet implements IteratorAggregate
```
Comment (nilmerg) [inline library/Icingadb/Util/PerfDataSet.php:10]:
```suggestion
class PerfDataSet implements IteratorAggregate
```

**Review by nilmerg — CHANGES_REQUESTED** [review]
Didn't test it until now. Noticed that you didn't use the same case for some class names in their respective file name. This works for you because your mac file system is case-insensitive, mine is not. I've added comments where the change is necessary, though you should use PHPStorm's refactor/rename functionality, as it automatically adjusts all usages as well.

I've also noticed some issues in the UI:

* If there's no critical value at all, the header for the column is missing
  Screenshot from 2021-04-21 11-47-05
* If there's a pie chart in the very first line, there should be a gap between it and the header
  Screenshot from 2021-04-21 11-47-19
* Since the values are all right-aligned now, I have some difficulty matching them with their respective label/pie-chart. Maybe show a faint line upon hover over a row or so.

**Review by nilmerg — CHANGES_REQUESTED** [review]
Looks mostly good to me now. We talked about how the padding looks to me in Chrome, though. Also, in mobile the table overflows, the labels should probably collapse?

Screenshot from 2021-04-22 08-48-55

**Comment by flourish86** [comment]
> Looks mostly good to me now. We talked about how the padding looks to me in Chrome, though. Also, in mobile the table overflows, the labels should probably collapse?
> 
> Screenshot from 2021-04-22 08-48-55

Chrome padding is fixed with 790cab22. Agree, that scrolling of the complete detail view is a problem. But wrapping the cell content could make the table layout fall apart completely and readability worse in tables with many rows.

I'd fix this with having only the table scroll horizontally, as it has been in an older version.

**Comment by flourish86** [comment]
> I'd fix this with having only the table scroll horizontally, as it has been in an older version.

See af0cb8aa

**Review by nilmerg — APPROVED** [review]

**Pull request merged**